Rubberduck.Core/UI/Refactorings/IRefactoringDialogFactory.cs

```diff
@@ -0,0 +1,18 @@
+﻿using Rubberduck.Refactorings;
+
+namespace Rubberduck.UI.Refactorings
+{
+    public interface IRefactoringDialogFactory
+    {
+        TDialog CreateDialog<TModel, TView, TViewModel, TDialog>(TModel model, TViewModel viewmodel)
+            where TModel : class
+            where TView : class, IRefactoringView, new()
+            where TViewModel : class, IRefactoringViewModel<TModel>
+            where TDialog : class, IRefactoringDialog<TModel, TView, TViewModel>;
+        void ReleaseDialog(IRefactoringDialog dialog);
```

Comment:
why are the generics here not specified out?

Reply:
The `CreateDialog` method returns `TDialog`, which must be a `IRefactoringDialog<>`, which is derived from non-generic `IRefactoringDialog` interface. [See here](https://github.com/rubberduck-vba/Rubberduck/pull/4072/files/7cd7fd05b902790dfea0835a05b7b95a37bd8418#diff-c3f4d72d96555085979e7bc9ad3a99edR11). Releasing the object isn't really concerned with the particulars of type parameters.

Reply:
BTW, one more background info that may help --- originally I had the factory with generics at the top but that put us in the awkward of needing to specifying all the type parameters in the refactoring class, meaning it's now tightly coupled which is no good. 

By moving the type parameters out of the interface and into the creation method, it becomes easier to substitute the appropriate type. For example, `RenameRefactoring` only need to specify `IRenamePresenter` interface to create it as [seen here](https://github.com/rubberduck-vba/Rubberduck/pull/4072/files/7cd7fd05b902790dfea0835a05b7b95a37bd8418#diff-3aef8b9d20776e2c4da2e947148823f6R85) and CW is able to locate the correct implementation to [this concrete presenter](https://github.com/rubberduck-vba/Rubberduck/pull/4072/files/7cd7fd05b902790dfea0835a05b7b95a37bd8418#diff-e6ab9c62b77591f8b71cf1d0c2071b94R7) without specifying the generics at all inside the refactoring class.

Reply:
Are you sure CW releases anything if the type of the parameter of the release method is not an exact match?

Reply:
Good question. I did not find anything for or against the idea, so I will have to test this myself. Failing that, [this blog post](http://kozmic.net/2009/12/23/castle-typed-factory-facility-reborn/) suggests in the comments that I can make the factory disposable which will release all the objects created by that factory. That might be simpler and can be implemented in the base class.

Reply:
If you go that route, just make sure not to forget to make the factory transient.

Reply:
As I didn't know how I'd test that, I also [asked on Stack Overflow](https://stackoverflow.com/questions/50827322/different-parameters-type-for-typed-factory-in-castle-windsor).